Ballerina-Training/Ballerina-Forum-Application/back-end/bal-forum/main.bal

```diff
@@ -0,0 +1,222 @@
+        if userDb is UserDb {
+            User user = check transformUserFromDatabase(userDb);
+            return user;
+        }
+
+        NotFound userNotFound = {
+            body: {
+                error_message: "User not found"
+            }
+        };
+        return userNotFound;
+    }
+
+    resource function post login(UserLogin userLogin) returns LoginOk|Unauthorized {
+        User|error user = forumDbClient->queryRow(`
+            SELECT id, name, email FROM users WHERE name = ${userLogin.name} && password = ${userLogin.password}
+        `);
+        if user is User {
+            LoginOk loginOk = {
+                body: {
+                    message: "Login successfully",
+                    user: user
+                }
+            };
+            return loginOk;
+        }
+
+        Unauthorized unauthorized = {
+            body: {
+                error_message: "Invalid credentials"
+            }
+        };
+        return unauthorized;
+    }
+
+    resource function post users/[string id]/posts(NewPost newPost) returns Created|Forbidden|error {
+        Sentiment sentiment = check sentimentClient->/api/sentiment.post({
+            text: newPost.description
+        });
+        if sentiment.label == "neg" {
+            Forbidden forbidden = {
+                body: {
+                    error_message: "Post rejected due to negative sentiment"
+                }
+            };
+            return forbidden;
+        }
+
+        UserDb userDb = check forumDbClient->queryRow(`
+            SELECT name, email, id, subscribtions FROM users WHERE id = ${id}
+        `);
+
+        User user = check transformUserFromDatabase(userDb);
+        Post post = check createPostFromNewPost(newPost, user.name);
+
+        _ = check forumDbClient->execute(`
+            INSERT INTO posts VALUES (${post.id}, ${post.title}, ${post.description}, ${post.username}, ${post.likes.toJsonString()}, ${post.comments.toJsonString()}, ${post.postedAt})
+        `);
+
+        if user.subscribtions.indexOf(post.id) is () {
+            user.subscribtions.push(post.id);
+            _ = check forumDbClient->execute(`
+                UPDATE users SET subscribtions = ${user.subscribtions.toJsonString()} WHERE id = ${id}
+            `);
+        }
+
+        Created created = {
+            body: {
+                message: "Post created successfully!"
+            }
+        };
+        return created;
+    }
+
+    resource function get posts() returns Post[]|error {
+        stream<PostDb, sql:Error?> postStream = forumDbClient->query(`SELECT * FROM posts`);
+        Post[] posts = check from var post in postStream
+            select check transformPostFromDatabase(post);
+        return posts;
+    }
+
+    resource function post posts/[string id]/like(Like like) returns Ok|NotFound|Conflict|error {
+        PostDb|error postDb = forumDbClient->queryRow(`SELECT * FROM posts WHERE id = ${id}`);
+        if postDb is PostDb {
```

Comment:
Can't we swap the check and return on error instead? Will reduce nesting for the core logic.

Reply:
+1

Reply:
Fixed via [1eae7ab](https://github.com/ballerina-guides/training-content/pull/12/commits/1eae7abbc8d1d6bdd3a677b4fac348875ee53fb7)